Getting InvalidRequestException during fetch with twmb/franz-go/pkg/kmsg v1.11.0

Hello @twmb, hope you are doing great!

I just ran our integration tests with twmb/franz-go/pkg/kmsg v1.11.0 and got `InvalidRequestException` in Kafka logs during fetch, and test suite failed:

```
[2025-04-11 19:50:03,285] ERROR Exception while processing request from 172.18.0.3:29092-192.168.65.1:26237-32 (kafka.network.Processor)
org.apache.kafka.common.errors.InvalidRequestException: Error getting request for apiKey: FETCH, apiVersion: 16, connectionId: 172.18.0.3:29092-192.168.65.1:26237-32, listenerName: ListenerName(PLAINTEXT_HOST), principal: User:ANONYMOUS
Caused by: java.lang.RuntimeException: Tag 0 is not valid for version 16
	at org.apache.kafka.common.message.FetchRequestData$FetchPartition.read(FetchRequestData.java:1569)
	at org.apache.kafka.common.message.FetchRequestData$FetchPartition.<init>(FetchRequestData.java:1509)
	at org.apache.kafka.common.message.FetchRequestData$FetchTopic.read(FetchRequestData.java:1188)
	at org.apache.kafka.common.message.FetchRequestData$FetchTopic.<init>(FetchRequestData.java:1129)
	at org.apache.kafka.common.message.FetchRequestData.read(FetchRequestData.java:325)
	at org.apache.kafka.common.message.FetchRequestData.<init>(FetchRequestData.java:245)
	at org.apache.kafka.common.requests.FetchRequest.parse(FetchRequest.java:467)
	at org.apache.kafka.common.requests.AbstractRequest.doParseRequest(AbstractRequest.java:180)
	at org.apache.kafka.common.requests.AbstractRequest.parseRequest(AbstractRequest.java:172)
	at org.apache.kafka.common.requests.RequestContext.parseRequest(RequestContext.java:120)
	at kafka.network.RequestChannel$Request.<init>(RequestChannel.scala:107)
	at kafka.network.Processor.$anonfun$processCompletedReceives$1(SocketServer.scala:1151)
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(Unknown Source)
	at kafka.network.Processor.processCompletedReceives(SocketServer.scala:1129)
	at kafka.network.Processor.run(SocketServer.scala:1015)
	at java.base/java.lang.Thread.run(Unknown Source)
```

I tried twmb/franz-go/pkg/kmsg v1.10.0 - this one does not compile, and with twmb/franz-go/pkg/kmsg v1.9.0 tests work fine.

Also, this only happen with latest confluentinc/cp-kafka:7.9.0 (I tried 7.8.2 – it worked well)

Maybe you have ideas what's going on here? Not very clear is it a regression in franz-go, or in Kafka itself.

## Reproducing test case:

Make `docker-compose.yml`:

```yaml
version: "3.8"

services:
  kafka:
    image: confluentinc/cp-kafka:7.9.0
    depends_on:
      - zookeeper
    ports:
      - "29092:29092"
    healthcheck:
      test: ["CMD", "kafka-topics", "--list", "--bootstrap-server", "localhost:9092"]
      interval: 2s
      timeout: 5s
      retries: 10
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092,PLAINTEXT_HOST://localhost:29092
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
      KAFKA_INTER_BROKER_LISTENER_NAME: PLAINTEXT
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      KAFKA_TRANSACTION_STATE_LOG_MIN_ISR: 1
      KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 1
```

Then `docker compose up`.

`franz-go` dependencies:

```
github.com/twmb/franz-go v1.18.1
github.com/twmb/franz-go/pkg/kadm v1.15.0
github.com/twmb/franz-go/pkg/kmsg v1.11.0
```

Test case:

```go
package consumer

import (
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/twmb/franz-go/pkg/kadm"
	"github.com/twmb/franz-go/pkg/kgo"

	"github.com/google/uuid"
	"github.com/stretchr/testify/require"
)

const (
	testKafkaBrokerURL = "localhost:29092"
)

func makeTestTopic(topicName string, numPartitions int32, replicationFactor int16) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	cl, err := kgo.NewClient(kgo.SeedBrokers(testKafkaBrokerURL))
	if err != nil {
		return fmt.Errorf("failed to create Kafka client: %w", err)
	}
	defer cl.Close()

	client := kadm.NewClient(cl)
	defer client.Close()

	// Create the topic
	resp, err := client.CreateTopics(ctx, numPartitions, replicationFactor, nil, topicName)
	if err != nil {
		return fmt.Errorf("failed to create topic: %w", err)
	}

	for _, topic := range resp.Sorted() {
		if topic.Err != nil {
			if strings.Contains(topic.Err.Error(), "TOPIC_ALREADY_EXISTS") {
				continue
			}
			return fmt.Errorf("failed to create topic '%s': %v", topic.Topic, topic.Err)
		}
	}
	return nil
}

func produceMessage(topic string, message []byte) error {
	// Create a new client
	client, err := kgo.NewClient(kgo.SeedBrokers(testKafkaBrokerURL))
	if err != nil {
		return fmt.Errorf("failed to create Kafka client: %w", err)
	}
	defer client.Close()

	// Produce a message
	err = client.ProduceSync(context.Background(), &kgo.Record{
		Topic:     topic,
		Partition: 0,
		Value:     message,
	}).FirstErr()
	if err != nil {
		return fmt.Errorf("failed to produce message: %w", err)
	}
	return nil
}

func TestIssue(t *testing.T) {
	testKafkaTopic := "consumer_test_" + uuid.New().String()

	err := makeTestTopic(testKafkaTopic, 1, 1)
	require.NoError(t, err)

	err = produceMessage(testKafkaTopic, []byte("test message"))
	require.NoError(t, err)

	opts := []kgo.Opt{
		kgo.SeedBrokers("localhost:29092"),
		kgo.ConsumeTopics(testKafkaTopic),
		kgo.ConsumerGroup(uuid.New().String()),
		kgo.AutoCommitMarks(),
		kgo.BlockRebalanceOnPoll(),
	}

	client, err := kgo.NewClient(opts...)
	require.NoError(t, err)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	fetches := client.PollRecords(ctx, 100)
	require.NoError(t, fetches.Err())
}
```

This fails (and exception above present in Kafka container logs):

```
--- FAIL: TestIssue (5.06s)
    kafka_broken_test.go:90:
        	Error Trace:	/Users/alexander.emelin/consuming/kafka_broken_test.go:90
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestIssue
```

While works with confluentinc/cp-kafka:7.8.2 or with github.com/twmb/franz-go/pkg/kmsg v1.9.0

--- a/generate/gen.go
+++ b/generate/gen.go
@@ -190,20 +190,19 @@ func (s Struct) WriteAppend(l *LineWriter) {
 	l.Write("if isFlexible {")
 	defer l.Write("}")
 
-	var tagsCanDefault bool
 	for i := 0; i < len(tags); i++ {
 		f, exists := tags[i]
 		if !exists {
 			die("saw %d tags, but did not see tag %d; expected monotonically increasing", len(tags), i)
 		}
-		if _, tagsCanDefault = f.Type.(Defaulter); tagsCanDefault {
-			break
+		if _, canDefault := f.Type.(Defaulter); !canDefault {
+			die("no Defaulter for %s", f.FieldName)
 		}
 	}
 
 	defer l.Write("dst = v.UnknownTags.AppendEach(dst)")
 
-	if tagsCanDefault {
+	if len(tags) > 0 {
 		l.Write("var toEncode []uint32")
 		for i := 0; i < len(tags); i++ {
 			f := tags[i]
@@ -241,13 +240,17 @@ func (s Struct) WriteAppend(l *LineWriter) {
 		defer l.Write("}")
 		defer l.Write("}")
 	} else {
-		l.Write("dst = kbin.AppendUvarint(dst, %d + uint32(v.UnknownTags.Len()))", len(tags))
+		// No defined tags, we add unknown only.
+		//
+		// We keep a 0+ at the beginning for now to not cause a large
+		// delta on the generated code (previous versions of the
+		// generator did something slightly different here). This can
+		// be removed someday.
+		l.Write("dst = kbin.AppendUvarint(dst, 0+uint32(v.UnknownTags.Len()))")
 	}
 
 	for i := 0; i < len(tags); i++ {
-		if tagsCanDefault {
-			l.Write("case %d:", i)
-		}
+		l.Write("case %d:", i)
 		f := tags[i]
 
 		l.Write("{")

--- a/generate/main.go
+++ b/generate/main.go
@@ -375,6 +375,19 @@ func (s Struct) GetTypeDefault() any {
 	return fmt.Sprintf("(func() %[1]s { var v %[1]s; v.Default(); return v })() ", s.Name)
 }
 
+func (u Uuid) SetDefault(string) Type {
+	die("cannot set default on a uuid")
+	return u
+}
+
+func (Uuid) GetDefault() (any, bool) {
+	return "", false // no GetDefault
+}
+
+func (Uuid) GetTypeDefault() any {
+	return "[16]byte{}"
+}
+
 type FlexibleSetter interface {
 	AsFromFlexible() Type
 }

--- a/pkg/kmsg/generated.go
+++ b/pkg/kmsg/generated.go
@@ -4329,12 +4329,21 @@ func (v *FetchRequest) AppendTo(dst []byte) []byte {
 						dst = kbin.AppendInt32(dst, v)
 					}
 					if isFlexible {
-						dst = kbin.AppendUvarint(dst, 1+uint32(v.UnknownTags.Len()))
-						{
-							v := v.ReplicaDirectoryID
-							dst = kbin.AppendUvarint(dst, 0)
-							dst = kbin.AppendUvarint(dst, 16)
-							dst = kbin.AppendUuid(dst, v)
+						var toEncode []uint32
+						if v.ReplicaDirectoryID != [16]byte{} {
+							toEncode = append(toEncode, 0)
+						}
+						dst = kbin.AppendUvarint(dst, uint32(len(toEncode)+v.UnknownTags.Len()))
+						for _, tag := range toEncode {
+							switch tag {
+							case 0:
+								{
+									v := v.ReplicaDirectoryID
+									dst = kbin.AppendUvarint(dst, 0)
+									dst = kbin.AppendUvarint(dst, 16)
+									dst = kbin.AppendUuid(dst, v)
+								}
+							}
 						}
 						dst = v.UnknownTags.AppendEach(dst)
 					}
@@ -42175,12 +42184,21 @@ func (v *FetchSnapshotRequest) AppendTo(dst []byte) []byte {
 						dst = kbin.AppendInt64(dst, v)
 					}
 					if isFlexible {
-						dst = kbin.AppendUvarint(dst, 1+uint32(v.UnknownTags.Len()))
-						{
-							v := v.ReplicaDirectoryID
-							dst = kbin.AppendUvarint(dst, 0)
-							dst = kbin.AppendUvarint(dst, 16)
-							dst = kbin.AppendUuid(dst, v)
+						var toEncode []uint32
+						if v.ReplicaDirectoryID != [16]byte{} {
+							toEncode = append(toEncode, 0)
+						}
+						dst = kbin.AppendUvarint(dst, uint32(len(toEncode)+v.UnknownTags.Len()))
+						for _, tag := range toEncode {
+							switch tag {
+							case 0:
+								{
+									v := v.ReplicaDirectoryID
+									dst = kbin.AppendUvarint(dst, 0)
+									dst = kbin.AppendUvarint(dst, 16)
+									dst = kbin.AppendUuid(dst, v)
+								}
+							}
 						}
 						dst = v.UnknownTags.AppendEach(dst)
 					}

--- a/pkg/kmsg/go.mod
+++ b/pkg/kmsg/go.mod
@@ -3,3 +3,8 @@ module github.com/twmb/franz-go/pkg/kmsg
 go 1.21
 
 toolchain go1.22.0
+
+retract (
+	v1.11.0 // This version erroneously always encoded tagged uuid fields, which failed on any Kafka version that did not support the field
+	v1.10.0 // This version failed to compile; I must have accidentally git-broke my fixed commit before pushing & merging
+)

